File: androidApp/src/main/java/com/mirego/kmp/boilerplate/views/ExampleView.kt
package com.mirego.kmp.boilerplate.views

import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.tooling.preview.Preview
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.mirego.kmp.boilerplate.previews.PreviewContext
import com.mirego.kmp.boilerplate.viewmodels.example.ExampleViewModel

@Composable
fun ExampleView(viewModel: ExampleViewModel) {
    val state: ExampleViewModel.State by viewModel.state.collectAsStateWithLifecycle()

    Text(text = state.greeting)
}

@Preview(showSystemUi = true)
@Composable
fun PreviewExampleView() {
    PreviewContext { viewModelFactory ->
        ExampleView(
            viewModel = viewModelFactory.example()
        )
    }
}

File: ios/iosApp/Views/ExampleView.swift
import Shared
import SwiftUI

struct ExampleView: View {

    private let viewModel: ExampleViewModel

    @State private var state: ExampleViewModelState

    init(viewModel: ExampleViewModel) {
        self.viewModel = viewModel
        self.state = viewModel.state.value
    }

    var body: some View {
        VStack {
            Text(state.greeting)
        }
        .task {
            await withTaskCancellationHandler {
                for await state in viewModel.state {
                    self.state = state
                }
            } onCancel: {
                viewModel.cancel()
            }
        }
    }
}

#Preview {
    PreviewContext { viewModelFactory in
        ExampleView(
            viewModel: viewModelFactory.example()
        )
    }
}
